server/api.go

```diff
@@ -0,0 +1,1024 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"runtime/debug"
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/larkox/mattermost-plugin-badges/badgesmodel"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+// HTTPHandlerFuncWithUser is http.HandleFunc but userID is already exported
+type HTTPHandlerFuncWithUser func(w http.ResponseWriter, r *http.Request, userID string)
+
+// ResponseType indicates type of response returned by api
+type ResponseType string
+
+const (
+	// ResponseTypeJSON indicates that response type is json
+	ResponseTypeJSON ResponseType = "JSON_RESPONSE"
+	// ResponseTypePlain indicates that response type is text plain
+	ResponseTypePlain ResponseType = "TEXT_RESPONSE"
+	// ResponseTypeDialog indicates that response type is a dialog response
+	ResponseTypeDialog ResponseType = "DIALOG"
+)
+
+type APIErrorResponse struct {
+	ID         string `json:"id"`
+	Message    string `json:"message"`
+	StatusCode int    `json:"status_code"`
+}
+
+func (p *Plugin) initializeAPI() {
+	p.router = mux.NewRouter()
+	p.router.Use(p.withRecovery)
+
+	apiRouter := p.router.PathPrefix("/api/v1").Subrouter()
+	pluginAPIRouter := p.router.PathPrefix(badgesmodel.PluginAPIPath).Subrouter()
+	autocompleteRouter := p.router.PathPrefix(AutocompletePath).Subrouter()
+	dialogRouter := p.router.PathPrefix(DialogPath).Subrouter()
+
+	apiRouter.HandleFunc("/getUserBadges/{userID}", p.extractUserMiddleWare(p.getUserBadges, ResponseTypeJSON)).Methods(http.MethodGet)
+	apiRouter.HandleFunc("/getBadgeDetails/{badgeID}", p.extractUserMiddleWare(p.getBadgeDetails, ResponseTypeJSON)).Methods(http.MethodGet)
+	apiRouter.HandleFunc("/getAllBadges", p.extractUserMiddleWare(p.getAllBadges, ResponseTypeJSON)).Methods(http.MethodGet)
+
+	pluginAPIRouter.HandleFunc(badgesmodel.PluginAPIPathEnsure, checkPluginRequest(p.ensureBadges)).Methods(http.MethodPost)
+	pluginAPIRouter.HandleFunc(badgesmodel.PluginAPIPathGrant, checkPluginRequest(p.grantBadge)).Methods(http.MethodPost)
+
+	autocompleteRouter.HandleFunc(AutocompletePathBadgeSuggestions, p.extractUserMiddleWare(p.getBadgeSuggestions, ResponseTypeJSON)).Methods(http.MethodGet)
```

Comment:
Hey, why not use something like `autocompletRouter.Use(p.extractUserMiddleware(ResponseTypeJSON))`

This is only a suggestion, I don't know if you prefer the explicitness of the extractUserMiddleware everywhere.

Reply:
For the time being I will leave it like that. One of the things we are doing is passing down the user to the function, and it does not seem easy to do with `router.Use(...)`.

Reply:
Actually is not hard at all, you need to have a wrapping function, something like:

```go
func extractUserMiddlewareBuilder(responseType string) {
  return func (next http.Handler) http.Handler {
    return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
      ...
    }
  }
}
```

and then use it like:

```go
router.Use(extractUserMiddlewareBuilder(ResponseTypeJSON))
```

But I'm ok moving forward without that change, it is just a way to improve it, but not because now is necesarilly wrong.